`testers`: docs and move

roberth opened the pull request:
###### Description of changes

Closes #156785

Tidy up the testers a bit.

- move `invalidateFetcherByDrvHash` into place as expected from docs
- move `nixosTest` to `testers` too and document it in the manual

![image](https://user-images.githubusercontent.com/496447/166912106-9adafac1-b984-4be8-8815-1f36896a1ac3.png)
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
testers.invalidateFetcherByDrvHash: Move from top-level

roberth committed:
nixos/testing-python.nix: Set meta.mainProgram

roberth committed:
testers.nixosTest: Move from top-level and improve docs

roberth committed:
doc/testers: Mention nixosTests in nixosTest doc

ncfavier commented on doc/builders/testers.chapter.md:
```diff
@@ -80,3 +80,49 @@ tests.fetchgit = invalidateFetcherByDrvHash fetchgit {
+
+Notable attributes:
+
+ * `nodes`: the evaluated NixOS configurations. Useful for debugging and exploring the configuration.
```
Note that in https://github.com/NixOS/nixpkgs/pull/164660 I changed `nodes` from being a function that takes a qemu package to that function applied to `pkgs.qemu_test`. I didn't give it much thought back then, and I don't know what we actually want to expose here.

roberth replied:
I guess that change didn't invalidate the spirit of this doc. Did you have a specific reason to change it, or was it for implementation simplicity?

I think we could change it back when making the tests more flexible as I've proposed in #171396. The qemu package will be configurable, so if you want to apply it to another qemu, you can do so by defining the option value instead. I'm leaning towards leaving this as is for the time being.

ncfavier replied:
> Did you have a specific reason to change it, or was it for implementation simplicity?

It was just for simplicity.

ncfavier reviewed (commented):
LGTM, though there are still a couple uses of the top-level `nixosTest` alias left.

roberth committed:
treewide: nixosTest -> testers.nixosTest

roberth committed:
tests.testers.nixosTest-example: move from tests.nixos-functions.nixosTest-test

And improve the test a bit, to assert correct wiring of `pkgs`.

roberth committed:
Merge remote-tracking branch 'upstream/master' into testers

roberth committed:
testers.nixosTest: Remove redundant system.stateVersion = lib.trivial.release;

Already present in nixos/modules/testing/test-instrumentation.nix,
which is imported by the test framework.
